openssl-dynamic/src/main/c/tcn_lock_rw.cpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+#include "tcn_lock_rw.h"
+#include <shared_mutex>
+
+tcn_lock_rw_t tcn_lock_rw_new() {
+    auto *m = new std::shared_timed_mutex();
+    return (tcn_lock_rw_t) m;
+}
+
+void tcn_lock_rw_free(tcn_lock_rw_t lock) {
+    auto *m = (std::shared_timed_mutex *) lock;
+    delete m;
+}
+
+tcn_lock_writer_t tcn_lock_writer_acquire(tcn_lock_rw_t lock) {
+    auto *m = (std::shared_timed_mutex *) lock;
+    auto *writer_lock = new std::lock_guard<std::shared_timed_mutex>(*m);
```

Comment:
Did you consider using `std::unique_lock`? It is explicitly called out in the `std::shared_lock` docs (see below). Also the default behavior seems to be the same (constructor locks and destructor unlocks).

http://en.cppreference.com/w/cpp/thread/shared_lock
> Locking a shared_lock locks the associated shared mutex in shared mode (to lock it in exclusive mode, std::unique_lock can be used)